src/java.desktop/share/classes/javax/swing/border/EtchedBorder.java

```diff
@@ -131,22 +150,41 @@ public EtchedBorder(int etchType, Color highlight, Color shadow)    {
      * @param height the height of the painted border
      */
     public void paintBorder(Component c, Graphics g, int x, int y, int width, int height) {
-        int w = width;
-        int h = height;
+        // We remove any initial transforms to prevent rounding errors
+        // when drawing in non-integer scales
+        AffineTransform at = new AffineTransform();
+        Stroke oldStk = new BasicStroke();
+        int stkWidth = 1;
+        if (g instanceof Graphics2D) {
+            at = ((Graphics2D) g).getTransform();
```

Comment:
We more usually write
Graphics2D g2d = (Graphics2D)g

then just use g2d instead of repeated casting.

Reply:
This could even use pattern matching:
```java
        if (g instanceof Graphics2D g2d) {
            at = g2d.getTransform();
```

Reply:
I thought about that but (1) couldn't remember off-hand if it was still preview, (2) I was sure it would impede a backport ...

Reply:
[JEP 394: Pattern Matching for instanceof](https://openjdk.java.net/jeps/394) was delivered in JDK 16, I've seen it used a few times in JDK code.

It would impede backporting to JDK 11.